File: python3_solution_set1/859. Buddy Strings.py
##ss
class Solution:
    def buddyStrings(self, s: str, goal: str) -> bool:
        
        diff = 0
        conn = []
        if len(s)!=len(goal):
            return False
        
        for x in range(len(s)):
            if s[x]!=goal[x]:
                diff+=1
                conn.append([s[x],goal[x]])
                
             
        
        if diff in [1]:
            return False
        
        if diff == 0:
            c = Counter(s)
            for x in c:
                if c[x] >=2:
                    return True
                
            return False
        
        if diff > 2:
            return False
        
        elif diff == 2:
            if conn[0][0] == conn[1][1] and conn[0][1] == conn[1][0]:
                return True
            
            return False
        

File: go_solution_set1/859. Buddy Strings.go
func find_map(s string) map[string]int{
    f := make(map[string]int)
    for x:=0;x<len(s);x++{
        _,ok := f[string(s[x])]
        if ok{
            f[string(s[x])]+=1
        } else {
            f[string(s[x])] = 1
        }
    }
    return f
}
func buddyStrings(s string, goal string) bool {
    
    cnt := 0
    smap :=find_map(s)
    goalmap := find_map(goal)
    for x:=0; x<len(s);x++{
        if s[x]!=goal[x]{
            cnt++
        }
    }
    if !reflect.DeepEqual(smap,goalmap){
        return false
    }
    if cnt == 2{
        return true
    }
    if cnt == 0{
         for _,val := range smap{
             if val>=2{
                 return true
             }
         }
    }
    
    
    return false
}
